test/blackbox-tests/test-cases/cram/cram-action-shell-opt.t

```diff
@@ -0,0 +1,153 @@
+Check that Cram tests run without shell option specified
+
+  $ echo '(lang dune 3.10)' > dune-project
+
+  $ cat > dune <<EOF
+  > (rule
+  >  (target foo.cram.corrected)
+  >  (mode (promote (until-clean)))
+  >  (action (cram foo.cram)))
+  > EOF
+
+  $ cat > foo.cram <<EOF
+  >   $ echo "foo from foo.cram"
+  > EOF
+
+  $ dune build
+  $ cat foo.cram.corrected
+    $ echo "foo from foo.cram"
+    foo from foo.cram
+
+Prepare custom and mock shells for following tests
+
+  $ dune clean
+  $ cat > dune <<EOF
+  > (executables (names sh bash custom_shell))
+  > (rule (mode (promote (until-clean))) (action (copy %{dep:./sh.exe} sh)))
+  > (rule (mode (promote (until-clean))) (action (copy %{dep:./bash.exe} bash)))
+  > EOF
+
+  $ cat > sh.ml <<EOF
+  > print_endline "sh.ml"
+  > EOF
+
+  $ cat > bash.ml <<EOF
+  > print_endline "bash.ml"
+  > EOF
+
+  $ cat > custom_shell.ml <<EOF
+  > print_endline "custom_shell.ml"
+  > EOF
+
+  $ dune build
+  $ ls
+  _build
+  bash
+  bash.ml
+  custom_shell.ml
+  dune
+  dune-project
+  foo.cram
+  sh
+  sh.ml
+
+  $ ./sh
+  sh.ml
+  $ ./bash
+  bash.ml
+
+Check that shell option unset uses `env sh`
+
+  $ PATH=$PWD:$PATH
+  $ cat > dune <<EOF
+  > (rule
+  >  (target foo.cram.corrected)
+  >  (mode (promote (until-clean)))
+  >  (action (cram foo.cram)))
+  > EOF
+
+  $ cat > foo.cram <<EOF
+  >   $ echo "foo from foo.cram"
+  > EOF
+
+  $ dune build
+  sh.ml
+  $ cat foo.cram.corrected
+    $ echo "foo from foo.cram"
+    ***** UNREACHABLE *****
```

Comment:
Why is this and the rest unreachable?

Reply:
My guess is that when creating `foo.cram`, the `$` is interpreted by the parent shell. You can use `<< 'EOF'` line 69 to prevent this.

Reply:
These tests try to see whether the PATH environment variable is regarded by making our own `bash` and `sh` compiled from `bash.ml` and `sh.ml`. But since those are nowhere near complete shell implementation, they do not handle stdout redirection directives (like `echo hello > file.txt`), which cram_exec relies on to capture the output. Therefore the cram test executor will think the cram test crashed and include `***** UNREACHABLE *****` here.

We know our own mocked version of `bash` and `sh` are invoked here by looking at https://github.com/ocaml/dune/pull/8134/files/e2913128ca6046e24b2dbfa7255be40588ce58d1#diff-0484438795287a34ad56a4a530a816caea0a5480e23fffb006408f39591b35e6R74 and https://github.com/ocaml/dune/pull/8134/files/e2913128ca6046e24b2dbfa7255be40588ce58d1#diff-0484438795287a34ad56a4a530a816caea0a5480e23fffb006408f39591b35e6R95

Reply:
This is less than ideal but I didn't come up with other ways to perform this test.